core/src/sumeragi/mod.rs

```diff
@@ -127,22 +143,76 @@ impl Sumeragi {
             .internal
             .wsv
             .lock()
-            .expect("Failed to lock on `update_metrics`. Mutex poisoned");
+            .expect("Failed to lock on `wsv`. Mutex poisoned");
+
+        let metrics_guard = self
+            .metrics_mutex
+            .lock()
+            .expect("Failed to lock on `metrics`. Mutex poisoned");
+
+        let mut last_block_height_guard = self
+            .last_block_height
+            .lock()
+            .expect("Failed to lock on `last_block_height_when_updating_metrics`. Mutex poisoned");
+
+        let start_index = *last_block_height_guard;
+        {
+            let blocks_iter = wsv_guard.blocks();
+            let blocks_iter =
+                blocks_iter.skip(start_index.try_into().expect("Failed to cast to u32."));
+            for block in blocks_iter {
+                let block_txs_accepted = block.as_v1().transactions.len() as u64;
+                let block_txs_rejected = block.as_v1().rejected_transactions.len() as u64;
+
+                metrics_guard
+                    .txs
+                    .with_label_values(&["accepted"])
+                    .inc_by(block_txs_accepted);
+                metrics_guard
+                    .txs
+                    .with_label_values(&["rejected"])
+                    .inc_by(block_txs_rejected);
+                metrics_guard
+                    .txs
+                    .with_label_values(&["total"])
+                    .inc_by(block_txs_accepted + block_txs_rejected);
+                metrics_guard.block_height.inc();
+            }
+            *last_block_height_guard = wsv_guard.height();
+        }
+
+        metrics_guard.domains.set(wsv_guard.domains().len() as u64);
```

Comment:
I'd prefer this be computed on-demand in response to a query.

Reply:
It is. `update_metrics` is called when there is a query.